### PR TITLE
OT275 40: Servicio reutilizable para subida de imágenes

### PR DESCRIPTION
**Pull request by Sieglind**
# [OT275-40](https://alkemy-labs.atlassian.net/browse/OT275-40)

Created ImageDelegate class  and IUploadImage interface to allow the upload of images to the AWS S3 Service

### HOW HAS THIS BEEN TESTED?

I've implemented a bit of code on the PSVM to upload an image once the project is running

- [ ] Postman.
- [ ] Unit test.
- [ ] Integration test.
- [ ] No testing required (only applies for tech task).

**Test Configuration**:

* Used personal AWS S3 credentials due to unexpected behavior from @Value annotation in "AwsConfig" class

**Screenshots**:
Implemented this code to test the image upload:
image

The method "upload" returns this URL
image

The image is now available in the server and the URL matches the one returned to the LOG
image

* ....

### CHECKLIST

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have added the new resource in the Postman Collection file.

### CODE REVIEW

File: src/main/java/com/alkemy/ong/infrastructure/config/aws/delegate/ImageDelegate.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+
+@AllArgsConstructor
+@Service
+public class ImageDelegate implements IUploadImage {
+
+  private final AwsConfig awsConfig;
+
+  private static final String HTTPS_PART = "https://";
+  private static final String AMAZON_PART = ".s3.amazonaws.com/";
```
Comment (mkain0) [inline src/main/java/com/alkemy/ong/infrastructure/config/aws/delegate/ImageDelegate.java:19]: READ/REVIEW DOCUMENTATION: Esto valores hardcodeados funcionan por un tiempo mientras AWS no se le ocurra modificar la URL. Para esto el SDK de AWS tiene métodos que devuelven la url de un object.

File: src/main/java/com/alkemy/ong/infrastructure/config/aws/delegate/ImageDelegate.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+  public String upload(InputStream content, String contentType, String fileName) {
+    ObjectMetadata metadata = new ObjectMetadata();
+    metadata.setContentType(contentType);
+    AmazonS3 awsClient = awsConfig.generateS3client();
+    awsClient.putObject(
+        awsConfig.getBucketName(),
+        fileName,
+        content,
+        metadata);
```
Comment (mkain0) [inline src/main/java/com/alkemy/ong/infrastructure/config/aws/delegate/ImageDelegate.java:29]: Falta setear el ACL.

File: src/main/java/com/alkemy/ong/application/service/delegate/IUploadImage.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package com.alkemy.ong.application.service.delegate;
+
+import java.io.InputStream;
+
+public interface IUploadImage {
+
+  String upload(InputStream inputStream, String contentType, String fileName);
```
Comment (mkain0) [inline src/main/java/com/alkemy/ong/application/service/delegate/IUploadImage.java:7]: SOLID: Todos estos parámetros representan algo que se puede abstraer.

File: src/main/java/com/alkemy/ong/infrastructure/config/aws/delegate/ImageDelegate.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
+package com.alkemy.ong.infrastructure.config.aws.delegate;
```
Comment (mkain0) [inline src/main/java/com/alkemy/ong/infrastructure/config/aws/delegate/ImageDelegate.java:1]: TRIVIAL: podría estar directo en el package aws. No es necesario crear un package extra ya que solo hay dos clases.

**Review by mkain0 — CHANGES_REQUESTED** [review]

File: src/main/java/com/alkemy/ong/application/service/delegate/IUploadImage.java (hunk 3 above)
Comment (Sieglind) [reply src/main/java/com/alkemy/ong/application/service/delegate/IUploadImage.java:7]: Debería crear una clase de dominio que represente el contenido de una petición de upload?

Comment (mkain0) [reply src/main/java/com/alkemy/ong/application/service/delegate/IUploadImage.java:7]: Una interfaz. Revisa el repo de la meeting de abstracción.

**Comment by Sieglind** [comment]
Actualize las evidencias del funcionamiento del código

**Review by mkain0 — APPROVED** [review]

**Review by mkain0 — APPROVED** [review]

**Pull request merged**